### PR TITLE
Initial Implementation

### PR DESCRIPTION
**Pull request by muvaf**
Implementing https://github.com/crossplaneio/crossplane/pull/1110

Implementation of the generic resource pack reconciler.

TODOs:
- [x] Variable references from CR to child resources. This will be done via [custom transformer](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/crd/README.md#adding-a-custom-resource) with `varReference`. Controller will dump the CR before Kustomization and generated resources will have the correct values taken from CR instance.
- [x] Default resource class handling.
- [x] Polishing the code
- [ ] Unit tests.

### CODE REVIEW

File: pkg/controllers/configuration_stack_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+	if meta.WasDeleted(cr) {
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	childResources, err := r.kustomizeOperation.RunKustomize(cr)
+	if err != nil {
+		return ctrl.Result{RequeueAfter: r.shortWait}, errors.Wrap(err, "kustomize operation failed")
+	}
+	childResources, err = r.childResourcePatcher.Patch(cr, childResources)
```
Comment (suskin) [inline pkg/controllers/configuration_stack_reconciler.go]: At the top level this seems pretty straightforward, but I am wondering whether you have some samples of objects that you've been using for testing, so I can get a better feel for what the objects actually look like.

If I am reading the minimal gcp implementation correctly, it looks as though the patcher can do pretty much whatever the author wants?

File: pkg/controllers/configuration_stack_reconciler.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	if err != nil {
+		return ctrl.Result{RequeueAfter: r.shortWait}, errors.Wrap(err, "child resource patchers failed")
+	}
+	for _, o := range childResources {
+		if err := Apply(ctx, r.kube, o); err != nil {
```
Comment (suskin) [inline pkg/controllers/configuration_stack_reconciler.go]: Do we need to worry about the delete case?

File: pkg/operations/kustomize.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	kustomizer := krusty.MakeKustomizer(filesys.MakeFsOnDisk(), krusty.MakeDefaultOptions())
+	resMap, err := kustomizer.Run(tempDir)
+	if err != nil {
+		return nil, err
+	}
+	var objects []resource.ChildResource
+	for _, res := range resMap.Resources() {
+		// ParentResource is written to kustomization directory only to be used
+		// for value retrieval.
+		if res.GetKind() == cr.GetObjectKind().GroupVersionKind().Kind {
```
Comment (suskin) [inline pkg/operations/kustomize.go]: What if it generates objects of the same kind, but of a different group?

File: pkg/operations/kustomize.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	Patcher resource.KustomizationPatcherChain
+}
+
+// RunKustomize creates a temporary directory that contains ParentResource YAML
+// and main kustomization.yaml file that refers to the original resources folder.
+// The temporary folder is deleted after the function exits.
+func (o *KustomizeOperation) RunKustomize(cr resource.ParentResource) ([]resource.ChildResource, error) {
+	// Kustomize does not work with symlinked paths, so, we're using its own
+	// temp directory generation function.
+	tempConfirmedDir, err := filesys.NewTmpConfirmedDir()
```
Comment (suskin) [inline pkg/operations/kustomize.go]: Are there any security or performance implications to using the filesystem?

File: pkg/operations/kustomize.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	// filesystem. It should be given as absolute path.
+	ResourcePath string
+
+	Patcher resource.KustomizationPatcherChain
+}
+
+// RunKustomize creates a temporary directory that contains ParentResource YAML
+// and main kustomization.yaml file that refers to the original resources folder.
+// The temporary folder is deleted after the function exits.
+func (o *KustomizeOperation) RunKustomize(cr resource.ParentResource) ([]resource.ChildResource, error) {
```
Comment (suskin) [inline pkg/operations/kustomize.go]: This part seems like it'll be useful for when we're wanting Kustomize-like functionality in template stacks :pancakes:

**Review by suskin — APPROVED** [review]
Seems pretty reasonable to me :squid: - my main questions at this point are about what this looks and feels like for the user. But, looking at the [most recent design doc](https://github.com/crossplaneio/crossplane/pull/1110/files#diff-e31b554af7a563b612944eb964cbd549R146), it seems there are already some samples in there.

I am approving this, because the approach of "add a simple thing and iterate on it" seems like a good approach to me, and this looks fine for fulfilling that goal.

File: pkg/controllers/configuration_stack_reconciler.go (hunk 1 above)
Comment (muvaf) [reply pkg/controllers/configuration_stack_reconciler.go]: > If I am reading the minimal gcp implementation correctly, it looks as though the patcher can do pretty much whatever the author wants?

Yes, exactly. They can even do some filtering and return less objects. Here you can find the default patcher objects https://github.com/muvaf/configuration-stacks/pull/1/files#diff-e56296f966d57a46eebe6c20eb64a1ac

File: pkg/controllers/configuration_stack_reconciler.go (hunk 2 above)
Comment (muvaf) [reply pkg/controllers/configuration_stack_reconciler.go]: All child resources have owner reference to the parent resource. If you go a few lines up, you'll see that if the parent resource is deleted we return early and let the Kubernetes take care of cleaning.

File: pkg/operations/kustomize.go (hunk 3 above)
Comment (muvaf) [reply pkg/operations/kustomize.go]: Right, I guess I opt for kind just for easiness. Fixed.

File: pkg/operations/kustomize.go (hunk 4 above)
Comment (muvaf) [reply pkg/operations/kustomize.go]: Not that I'm aware of. I mean we already use it in kustomize calls to read the `resources` folder. I was worried about the leaks in the case of errors but in `defer` statement we clean everything and if the program crashes, it means the container will be restarted and temp folders get deleted anyway. So, we should be safe as far as I see.

File: pkg/controllers/resourcepack_reconciler.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+	errGetResource           = "could not get the custom resource"
+	errTemplatingOperation   = "templating operation failed"
+	errChildResourcePatchers = "child resource patchers failed"
+	errApply                 = "apply failed"
+)
+
+type ResurcePackReconcilerOption func(*ResurcePackReconciler)
+
+func AdditionalChildResourcePatcher(op ...resource.ChildResourcePatcher) ResurcePackReconcilerOption {
+	return func(reconciler *ResurcePackReconciler) {
```
Comment (negz) [inline pkg/controllers/resourcepack_reconciler.go]: There's a typo in `ResurcePackReconciler` - I think it should be `ResourcePackReconciler`.

File: pkg/controllers/resourcepack_reconciler.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+func Apply(ctx context.Context, kube client.Client, o resource.ChildResource) error {
+	existing := o.DeepCopyObject().(resource.ChildResource)
+	err := kube.Get(ctx, types.NamespacedName{Name: o.GetName(), Namespace: o.GetNamespace()}, existing)
+	if kerrors.IsNotFound(err) {
+		return kube.Create(ctx, o)
+	}
+	if err != nil {
+		return err
+	}
+	o.SetResourceVersion(existing.GetResourceVersion())
```
Comment (negz) [inline pkg/controllers/resourcepack_reconciler.go:171]: I suggest adding a comment explaining why we're doing this - it's not obvious to me. Will `MergeFrom` not update the resource version?

File: pkg/operations/kustomize/kustomize.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+	// filesystem. It should be given as absolute path.
+	ResourcePath string
+
+	// Patcher contains the modifications that you'd like to make to
+	// the overlay Kustomization object before calling kustomize.
+	Patcher KustomizationPatcherChain
+}
+
+func (o *KustomizeEngine) Run(cr resource.ParentResource) ([]resource.ChildResource, error) {
+	tmpl, err := ioutil.ReadFile(fmt.Sprintf("%s/%s", o.ResourcePath, templateFileName))
```
Comment (negz) [inline pkg/operations/kustomize/kustomize.go]: Would there be value in using https://github.com/spf13/afero to abstract the need for a real filesystem (e.g. during tests)?

Also, https://golang.org/pkg/path/filepath/#Join is generally preferable to using `fmt.Sprintf` to form file paths.

**Review by negz — APPROVED** [review]

File: pkg/operations/kustomize/kustomize.go (hunk 8 above)
Comment (muvaf) [reply pkg/operations/kustomize/kustomize.go]: Thanks for `Join` suggestion!

I'm not sure whether we really need afero at this point. I'll integrate it once we have more uses for this reconciler in Crossplane and want to write some tests.

**Pull request merged**